llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -447,6 +447,19 @@ bool SPIRVInstructionSelector::spvSelect(Register ResVReg,
     return selectUnOp(ResVReg, ResType, I, SPIRV::OpBitcast);
   case TargetOpcode::G_ADDRSPACE_CAST:
     return selectAddrSpaceCast(ResVReg, ResType, I);
+  case TargetOpcode::G_PTR_ADD: {
```

Comment:
Quite counter-intuitive mapping `G_PTR_ADD`->`OpSpecConstantOp`, how these `G_PTR_ADD` are produced in the first place?

Reply:
Perhaps, it looks strange. The only case when we have G_PTR_ADD is a result of the translation of this original code:
```
@a_var = addrspace(1) global [2 x i8] c"\01\01", align 1
@p_var = addrspace(1) global i8 addrspace(1)* getelementptr inbounds ([2 x i8], [2 x i8] addrspace(1)* @a_var, i32 0, i64 1), align 8
```
After the intrinsic insertion we get this:
```
  %2 = call i8 @llvm.spv.track.constant.i8.i8(i8 1, metadata i8 1)
  %3 = call i8 @llvm.spv.track.constant.i8.i8(i8 1, metadata i8 1)
  %4 = call i32 (...) @llvm.spv.const.composite(i8 %2, i8 %3)
  %5 = call i32 @llvm.spv.track.constant.i32.i32(i32 %4, metadata [2 x i8] c"\01\01")
  call void @llvm.spv.assign.type.i32(i32 %5, metadata [2 x i8] zeroinitializer)
  call void @llvm.spv.init.global.p1a2i8.i32([2 x i8] addrspace(1)* @a_var, i32 %5)
  call void @llvm.spv.assign.type.p1i8(i8 addrspace(1)* getelementptr inbounds ([2 x i8], [2 x i8] addrspace(1)* @a_var, i32 0, i64 1), metadata i8 addrspace(1)* getelementptr inbounds ([2 x i8], [2 x i8] addrspace(1)* @a_var, i32 0, i64 1))
  call void @llvm.spv.init.global.p1p1i8.p1i8(i8 addrspace(1)* addrspace(1)* @p_var, i8 addrspace(1)* getelementptr inbounds ([2 x i8], [2 x i8] addrspace(1)* @a_var, i32 0, i64 1))
```
And after the translator:
```
  %14:_(p1) = G_GLOBAL_VALUE @a_var
  %16:_(s64) = G_CONSTANT i64 1
  %15:_(p1) = G_PTR_ADD %14:_, %16:_(s64)
  %17:_(p1) = G_GLOBAL_VALUE @p_var
...
  G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.spv.assign.type), %13:_(s32), <0x560682267988>
  G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.spv.init.global), %14:_(p1), %13:_(s32)
  G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.spv.assign.type), %15:_(p1), <0x560682249328>
  G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.spv.init.global), %17:_(p1), %15:_(p1)
```
This global var constant initialization should be translated to OpSpecConstantOp. Once we discover other sources of G_PTR_ADD, we will have to implement better handler for this. Probably I need to add a comment about this.

Reply:
@zuban32, I have added a short comment about it.

Reply:
Oh, I see, we've introduced `spirv.gep` intrinsics exactly to get rid of those adds, now I remember. But why is it a specialized constant to be produced from such code? I thought the only way to have a specialized constants emitted from LLVM IR is to use a builtin which I can't find in the test in this PR

Reply:
It's OpSpecConstantOp, not OpSpecConstant. I think they can be mapped to symbolic relocations (with modification of symbol values) in some regular assemblers. The translator produces this code for the test:
```
    %a_var = OpVariable %_ptr_CrossWorkgroup__arr_uchar_ulong_2 CrossWorkgroup %11
       %17 = OpSpecConstantOp %_ptr_CrossWorkgroup_uchar InBoundsPtrAccessChain %a_var %uint_0 %ulong_1
    %p_var = OpVariable %_ptr_CrossWorkgroup__ptr_CrossWorkgroup_uchar CrossWorkgroup %17
```
This patch allows the backend to produce the similar code (currently in the wrong order and with some garbage after it):
```
        %23 = OpVariable %8 CrossWorkgroup %30  // p_var
        %24 = OpVariable %10 CrossWorkgroup %16 // a_var
...
        %30 = OpSpecConstantOp %5 70 %24 %18 %29
(then several unused duplicates of this OpSpecConstantOp)
```
I hope the topological sorting will solve the issue with ordering and some fixing in the "hoisting" of instructions will remove the unused OpSpecConstantOps. 

At the moment, I'm not sure if can we support this case in a different way without introducing any new entities in the backend.

Reply:
Oops, my bad. And yeah, that's a well-known problem of `SpecConstantOp` being the only way to express LLVM's global constexprs in SPIR-V even though that wasn't quite what they were originally designed for. 

Then I'd suggest checking if that `G_PTR_ADD` has only global or constant operands. I believe in future it may be useful to separate such selection from a general one since the spec allows quite many opcodes to be used in `SpecConstantOp` (and I suggest that what the community will require anyway).

And yes, the topological sort should help here.

Reply:
> Then I'd suggest checking if that G_PTR_ADD has only global or constant operands.
Thanks for the good suggestion.